File: src/com/reus/leetcode/RotateImage.java
package com.reus.leetcode;

/**
 * https://leetcode.com/problems/rotate-image
 */
public class RotateImage {
    public void rotate(int[][] matrix) {
        transpose(matrix);
        reflect(matrix);
    }

    private void transpose(int[][] matrix) {
        for (int i = 0; i < matrix.length; i++) {
            for (int j = i; j < matrix[i].length; j++) {
                int tmp = matrix[i][j];
                matrix[i][j] = matrix[j][i];
                matrix[j][i] = tmp;
            }
        }
    }

    private void reflect(int[][] matrix) {
        int n = matrix.length;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < matrix[i].length / 2; j++) {
                int tmp = matrix[i][j];
                matrix[i][j] = matrix[i][n - j - 1];
                matrix[i][n - j - 1] = tmp;
            }
        }
    }
}

File: src/com/reus/leetcode/google/arrays/RotateImage.kt
package com.reus.leetcode.google.arrays

/**
 * https://leetcode.com/problems/rotate-image/
 */
class RotateImage {
    fun rotate(matrix: Array<IntArray>): Unit {
        transpose(matrix)
        mirror(matrix)
    }

    private fun mirror(matrix: Array<IntArray>) {
        for (y in 0..matrix.size - 1) {
            for (x in 0..(matrix[y].size / 2) - 1) {
                val tmp = matrix[y][x]
                matrix[y][x] = matrix[y][matrix[y].size - 1 - x]
                matrix[y][matrix[y].size - 1 - x] = tmp
            }
        }
    }

    private fun transpose(matrix: Array<IntArray>) {
        for (y in 0..matrix.size - 1) {
            for (x in y + 1..matrix[y].size - 1) {
                val tmp = matrix[x][y]
                matrix[x][y] = matrix[y][x]
                matrix[y][x] = tmp
            }
        }
    }
}

fun main() {
    RotateImage().rotate(arrayOf(
        intArrayOf(1,2,3),
        intArrayOf(4,5,6),
        intArrayOf(7,8,9),
    ))
}